src/line_buffer.rs

```diff
@@ -0,0 +1,32 @@
+use std::io;
```

Comment:
In general there's probably no easy way to see the "diff" of these vs the corresponding `grep` files

Mostly I stripped out stuff that I decided we didn't want to support/target at least for now (eg binary-file detection (seems questionable to treat that as a "thing" when by definition we're only dealing with parse-able source code files?), the concept of "captures" (seems intriguing that we could actually do something similar as far as tree-sitter query "matches"/"captures" and maybe leverage some of the existing code but I wasn't clear on what that modeling would be so stripped it out for now), "replacing" (`ripgrep` doesn't edit your files, but it does have a "read-only" "search-and-replace" functionality, which seems like it might be nice to support something similar, but from what I could tell it strictly depends on captures in its current form so stripped it out))

And then made a few changes to better "match" our use case

I also stripped out all comments, I like the idea that the corresponding original `grep` code is "source-of-truth". So maybe we could add comments for some of the stuff we modified if we want to. And I guess if we end up wanting to expose any of the `grep`-based stuff as part of a crate-level API then we'd want to re-add the relevant doc comments (to be able to generate docs for those things)?